circuit/src/main/kotlin/com/slack/circuit/Circuit.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2022 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.circuit
+
+import androidx.compose.runtime.Immutable
+
+/**
+ * Circuit adapts [presenterFactories][PresenterFactory] to their corresponding renderable
+ * [uiFactories][ScreenViewFactory] using [screens][Screen]. Create instances using [the Builder]
+ * [Builder] and create new [Navigator] instances from this using [Circuit.navigator].
+ *
+ * For example,
+ *
+ * ```kotlin
+ * val circuit = Circuit.Builder()
+ *     .addUiFactory(AddFavoritesScreenViewFactory()
+ *     .addPresenterFactory(AddFavoritesPresenterFactory()
+ *     .build()
+ *
+ * val navigator = circuit.navigator(AddFavoritesScreen())
+ * ```
+ */
+@Immutable
+class Circuit private constructor(builder: Builder) {
+  private val uiFactories: List<ScreenViewFactory> = builder.uiFactories.toList()
+  private val presenterFactories: List<PresenterFactory> = builder.presenterFactories.toList()
+
+  fun presenter(screen: Screen, navigator: Navigator): Presenter<*, *>? {
+    return nextPresenter(null, screen, navigator)
```

Comment:
Instead of iterating over all presenter factories every time we need a new instance, did you consider using a Map and keying the factory on the screen class? We could pass the Screen class (or use a reified type) to `Circuit.Builder. addPresenterFactory` which could build the map.

Same thing should work for ui factories

Reply:
Moshi does do keys like this, but retrofit does not. It would have to be key'd by screens and they are not guaranteed to be immutable. They may also contain nontrivial amounts of data. Keying by screen class also doesn't work since they may contain different data. For that reason I think we should at least start with this, and we can always optimize later if we want.